src/reactive/consumer.py

```diff
@@ -56,3 +60,26 @@ def consume(mongodb_uri: str, queue_name: str) -> None:
                 job_details.run_url,
             )
             msg.ack()
+
+
+def _add_sigterm_handler(queue: SimpleQueue, signal_code: signal.Signals) -> None:
+    """Add a signal handler to clean up and exit.
+
+    Args:
+        queue: The queue to clean up.
+        signal_code: The signal code to handle.
+    """
+
+    def sigterm_handler(signal_code: int, _: FrameType | None) -> None:
+        """Handle a signal.
+
+        Requeue the message and exit.
+
+        Args:
+            signal_code: The signal code to handle.
+        """
+        logger.error("Signal '%s' received. Requeueing message.", signal.strsignal(signal_code))
+        queue.channel.basic_recover(requeue=True)
+        sys.exit(signal_code)
```

Comment:
If I remember correctly, the Python signal handling function can execute at any point during the execution of python bytecode. This means there's no guarantee that the internal state of `queue` even `logger` will remain consistent entering the single handling function, potentially leading to unexpected behavior.

Reply:
https://docs.python.org/3.10/library/signal.html#execution-of-python-signal-handlers indicates that the handler function is not called during bytecode execution (but "after" the next bytecode instruction). So I would not expect to have an inconsistent state for the `queue' or `logger' object.

Reply:
Bytecode is a very small unit of execution, which means that the event handler can be called during the execution of an internal method in queue or logger, and the internal method can be in the process of manipulating a series of internal states. This is the same reason why catching `KeyboardInterrupt` is usually not a good idea. https://docs.python.org/3/library/exceptions.html#KeyboardInterrupt

Reply:
You are right.
I changed the code to just exit. It seems that kombu will requeue the message in that case (tested it locally), due to https://github.com/celery/kombu/blob/46af1b94e061e098e0471efdffac46c1e8c567b0/kombu/transport/virtual/base.py#L274 and https://github.com/celery/kombu/blob/46af1b94e061e098e0471efdffac46c1e8c567b0/kombu/transport/virtual/base.py#L194

Reply:
By the way, is signal handling necessary only for the `consume` function? If so, should the signal handler be canceled before the function returns?

Reply:
Have updated the code to restore the default handler.